### PR TITLE
Adding plugin state listener

### PR DESCRIPTION
**Pull request by Eransho**
Handling plugin uninstall - https://codota.atlassian.net/browse/DEV-384?atlOrigin=eyJpIjoiOTNlYjc1ZWRjM2MyNGVhZTg5MTVmOWZiNWJhZjUxY2UiLCJwIjoiaiJ9

### CODE REVIEW

File: src/main/java/com/tabnine/lifecycle/TabNinePluginStateListener.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    @Override
+    public void install(@NotNull IdeaPluginDescriptor descriptor) {
+    }
+
+    @Override
+    public void uninstall(@NotNull IdeaPluginDescriptor descriptor) {
+        final PluginId pluginId = descriptor.getPluginId();
+        if (pluginId != null && pluginId.equals(Utils.getPluginId())) {
+            try {
+                TabNineProcess.startTabNine(true, Collections.singletonList("--uninstalling"));
```
Comment (amircodota) [inline src/main/java/com/tabnine/lifecycle/TabNinePluginStateListener.java:25]: just make sure we need --uninstalling and not --uninstalled

Comment (Eransho) [reply src/main/java/com/tabnine/lifecycle/TabNinePluginStateListener.java:25]: Yes, that's what we use in sublime plugin

File: src/main/java/com/tabnine/lifecycle/TabNinePluginStateListener.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import java.util.Collections;
+
+public class TabNinePluginStateListener implements PluginStateListener {
+
+    @Override
+    public void install(@NotNull IdeaPluginDescriptor descriptor) {
+    }
+
+    @Override
+    public void uninstall(@NotNull IdeaPluginDescriptor descriptor) {
```
Comment (shaiahr) [inline src/main/java/com/tabnine/lifecycle/TabNinePluginStateListener.java:21]: What about disable?

**Comment by Eransho** [comment]
> I think T9 supports 'hot deploy' (i.e. it can be enabled without restart). Will the listener be registered in that case?

How is that related to the uninstall event?

File: src/main/java/com/tabnine/lifecycle/TabNinePluginStateListener.java (hunk 2 above)
Comment (Eransho) [reply src/main/java/com/tabnine/lifecycle/TabNinePluginStateListener.java:21]: Disable will be handled separately. It requires changes in the binary as well

**Comment by shaiahr** [comment]
> > I think T9 supports 'hot deploy' (i.e. it can be enabled without restart). Will the listener be registered in that case?
> 
> How is that related to the uninstall event?

In Codota, we report both. Don't we want to do the same here?

**Comment by boaz-codota** [comment]
https://github.com/codota/tabnine-intellij/pull/54

**Pull request closed without merging**